spec.bs

```diff
@@ -427,10 +427,325 @@ The <dfn>standard pathname options</dfn> is an [=options=] [=struct=] with [=opt
+  1. Return |parser|'s [=constructor string parser/result=].
+</div>
+
+<div algorithm>
+To <dfn>change state</dfn> given a [=constructor string parser=] |parser|, a [=constructor string parser/state=] |state|, and a number |skip|:
+
+  1. If |state| is not "<a for="constructor string parser/state">`init`</a>", not "<a for="constructor string parser/state">`authority`</a>", and not "<a for="constructor string parser/state">`done`</a>", then set |parser|'s [=constructor string parser/result=][|state|] to the result of running [=make a component string=] given |parser|.
+  1. Run [=change state without setting the component=] given |parser|, |state|, and |skip|.
+</div>
+
+<div algorithm>
+To <dfn>change state without setting the component</dfn> give a [=constructor string parser=] |parser|, a [=constructor string parser/state=] |state|, and a number |skip|:
+
+  1. Set |parser|'s [=constructor string parser/state=] to |state|.
+  1. Set |parser|'s [=constructor string parser/component start=] to |parser|'s [=constructor string parser/token index=] + |skip|.
+  1. Increment |parser|'s [=constructor string parser/token index=] by |skip|.
+  1. Set |parser|'s [=constructor string parser/token increment=] to 0.
+</div>
+
+<div algorithm>
+To <dfn>rewind</dfn> given a [=constructor string parser=] |parser|:
+
+  1. Set |parser|'s [=constructor string parser/token index=] to |parser|'s [=constructor string parser/component start=].
+  1. Set |parser|'s [=constructor string parser/token increment=] to 0.
+</div>
+
+<div algorithm>
+To <dfn>rewind and set state</dfn> given a [=constructor string parser=] |parser| and a [=constructor string parser/state=] |state|:
+
+  1. Run [=rewind=] given |parser|.
+  1. Set |parser|'s [=constructor string parser/state=] to |state|.
+</div>
+
+<div algorithm>
+To <dfn>get a safe token</dfn> given a [=constructor string parser=] |parser| and a number |index|:
+
+  1. If |index| is less than |parser|'s [=constructor string parser/token list=]'s [=list/size=], then return |parser|'s [=constructor string parser/token list=][|index|].
+  1. [=Assert=]: |parser|'s [=constructor string parser/token list=]'s [=list/size=] is greater than or equal to 1.
+  1. Let |last index| be |parser|'s [=constructor string parser/token list=]'s [=list/size=] &minus; 1.
+  1. Let |token| be |parser|'s [=constructor string parser/token list=][|last index|].
+  1. [=Assert=]: |token|'s [=token/type=] is "<a for=token/type>`end`</a>".
+  1. Return |token|.
+</div>
+
+<div algorithm>
+To run <dfn>is a non-special pattern char</dfn> given a [=constructor string parser=] |parser|, a number |index|, and a string |value|:
+
+  1. Let |token| be the result of running [=get a safe token=] given |parser| and |index|.
+  1. If |token|'s [=token/value=] is not |value|, then return false.
+  1. If any of the following are true, then return true:
+    <ol>
```

Comment:
See my above nit about style, but even if you don't go that route, this should be a ul, not ol.